server/broadcast_test.go

```diff
@@ -519,3 +547,290 @@ func TestVerifyPixels(t *testing.T) {
+	assert.Equal(0, sv.calls)  // sanity check initial call count
+	assert.Len(bsm.sessMap, 1) // sanity check initial bsm map
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.NotNil(err)
+	assert.Equal(1, sv.calls)
+	assert.Equal(sv.err, err)
+	assert.Len(bsm.sessMap, 1) // No effect on map for now
+
+	// Check retryable errors, esp broadcast session removal from manager
+	sv.err = verification.ErrTampered
+	sv.retries = 10 // Do this to ensure we get a nil result
+	_, retryable := sv.err.(verification.Retryable)
+	assert.True(retryable)
+	verifier = newStubSegmentVerifier(sv)
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.NotNil(err)
+	assert.Equal(2, sv.calls)
+	assert.Equal(sv.err, err)
+	assert.Len(bsm.sessMap, 0)
+
+	// When retries are set to 0, results are returned anyway in case of error
+	// (and more generally, when attempts > retries)
+
+	// Check data gets re-saved into OS if retry succeeds with earlier params
+	sv = &stubVerifier{
+		retries: 1,
+		err:     verification.ErrTampered,
+		results: []verification.Results{
+			verification.Results{Score: 9},
+			verification.Results{Score: 1},
+		},
+	}
+	mem, ok := drivers.NewMemoryDriver(nil).NewSession("streamName").(*drivers.MemorySession)
+	assert.True(ok)
+	name, err := mem.SaveData("/rendition/seg/1", []byte("attempt1"))
+	assert.Nil(err)
+	assert.Equal([]byte("attempt1"), mem.GetData(name))
+	sess.BroadcasterOS = mem
+	verifier = newStubSegmentVerifier(sv)
+	URIs[0] = name
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.Equal(sv.err, err)
+
+	// Now "insert" 2nd attempt into OS
+	// and ensure 1st attempt is what remains after verification
+	_, err = mem.SaveData("/rendition/seg/1", []byte("attempt2"))
+	assert.Nil(err)
+	assert.Equal([]byte("attempt2"), mem.GetData(name))
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.Nil(err)
+	assert.Equal([]byte("attempt1"), mem.GetData(name))
+}
+
+func TestVerifier_HLSInsertion(t *testing.T) {
+	assert := assert.New(t)
+
+	// Ensure that the playlist has the correct URL after verification
+	// Following this sequence:
+	//   1. Verify seg1. Verification fails.
+	//   2. Verify seg2. Verification fails.
+	//   3. Max retries hit, assume false positives
+	//   4. Return seg1 as highest scoring result
+	//   5. Insert seg1 into playlist
+	mid := core.ManifestID("foo")
+	pl := &stubPlaylistManager{manifestID: mid}
+	drivers.S3BUCKET = "livepeer"
+	mem := drivers.NewS3Driver("", drivers.S3BUCKET, "", "").NewSession(string(mid))
+	assert.NotNil(mem)
+	genBcastSess := func(url string) *BroadcastSession {
+		segData := []*net.TranscodedSegmentData{
+			&net.TranscodedSegmentData{Url: url, Pixels: 100},
+		}
+
+		buf, err := proto.Marshal(&net.TranscodeResult{
+			Result: &net.TranscodeResult_Data{
+				Data: &net.TranscodeData{Segments: segData},
+			},
+		})
+		assert.Nil(err, fmt.Sprintf("Could not marshal results for %s", url))
+		ts, mux := stubTLSServer()
+		//defer ts.Close()
```

Comment:
Took different approach in https://github.com/livepeer/go-livepeer/pull/1096/commits/a908c149a95a16d4add837e2a3fe4bb6b0749716 . It's a bit of a kludge but the API is more satisfying. Do you still prefer the explicit approach?

Do you know how important it is to call `ts.Close()` for httptest servers?  It seems designed to incorporate server shutdowns into testing rather than any critical bookkeeping [1], but I'm having a hard time finding more detailed explanations of its intended usage.

[1] https://golang.org/src/net/http/httptest/server.go?s=5205:5229#L180

Reply:
> It's a bit of a kludge but the API is more satisfying. Do you still prefer the explicit approach?

I'm fine the approach in a908c14 - just felt like it would be good practice to be in the habit of closing/shutting things down when they're no longer used even in tests.

> Do you know how important it is to call ts.Close() for httptest servers? It seems designed to incorporate server shutdowns into testing rather than any critical bookkeeping

Not sure exactly how important it is - the [docs for NewServer()](https://golang.org/pkg/net/http/httptest/#NewServer) do mention that the caller should call Close() when finished. Perhaps its just to make sure that there are no idle connections lying around in tests.